projects/packages/autoloader/tests/php/lib/class-test-plugin-factory.php

```diff
@@ -136,6 +136,7 @@ public static function create_test_plugin( $is_mu_plugin, $version ) {
 			'classmap' => array( 'includes' ),
 			'psr-4'    => array(
 				self::TESTING_NAMESPACE => 'src',
+				self::TESTING_NAMESPACE . $namespace_version . '\\' => 'src' . DIRECTORY_SEPARATOR . $namespace_version,
```

Comment:
My only goal with this PR was to fix the tests that were marked incomplete by moving the unique classes into a unique namespace. You're right though in pointing out that it is taking advantage of a bug (I didn't even notice that 😅)

The case where a more-specific namespace autoload is removed in a subsequent release seems incredibly unlikely, though it would be good to document the defect. The other case where a non-overlapping namespace is removed also exists with class autoloading and is likely much harder to solve.

Reply:
I’m trying to understand what’s going on here. Let me know if any of this is incorrect or incomplete:

* The test environment has two `UniqueTestClass` classes, and they’re in different namespaces:
     * `\Automattic\Jetpack\AutoloaderTesting\Current`
     * `\Automattic\Jetpack\AutoloaderTesting\Currentmu`

* Currently, the test plugin factory doesn’t properly create the PSR-4 map. Rather than mapping each the namespaces above to the correct path, it just maps the `\Automattic\Jetpack\AutoloaderTesting` namespace to the plugin’s `src` path. The autoloader needs to select one of those paths for the `\Automattic\Jetpack\AutoloaderTesting` namespace, and it selects the mu plugin's path.

What is the bug in the autoloader?

Reply:
> Currently, the test plugin factory doesn’t properly create the PSR-4 map.

This is incorrect. It's expected behavior to map a base namespace prefix to a path and expect sub-namespaces to map to subdirectories. See changelogger for example, it [maps](https://github.com/Automattic/jetpack-changelogger/blob/db6b06561350dc6d12151b2ed6f3e59446f4f526/composer.json#L18) `Automattic\Jetpack\Changelogger\` to `src/` and expects that `Automattic\Jetpack\Changelogger\Plugins\SemanticVersioning` maps to `src/Plugins/SemverVersioning.php`.

Also of note is that PSR-4 allows mapping of a prefix to "at least one base directory", implying that multiple base directories are allowed. Composer's autoloader [explicitly supports this](https://getcomposer.org/doc/04-schema.md#psr-4).

> What is the bug in the autoloader?

What's the intended behavior of the autoloader?

1. It uses the most recent version of the module, but falls back to older versions for classes that aren't in the most recent version.
2. It selects the most recent version of the module, ignoring all other versions.

If it's supposed to be #​1, the autoloader should be able to find the class without needing the redundant PSR-4 namespace declaration for the sub-namespace. If it's supposed to be #​2, the redundant namespace declaration for the sub-namespace shouldn't work to find the class.

Reply:
Thank you for explaining this. I understand what’s going on now.

I agree with your original comment that this probably isn’t a good change to make. We should decide on the intended behavior, then implement tests that confirm that the intended behavior occurs.

Reply:
> Also of note is that PSR-4 allows mapping of a prefix to "at least one base directory", implying that multiple base directories are allowed.

This is [also supported by the Jetpack Autoloader](https://github.com/Automattic/jetpack-autoloader/blob/855fc907021827cf6e6de4c87e49a8de874bdcc2/src/class-version-loader.php#L143). I assumed incorrectly when I added the incomplete test messages, but realized when creating this PR it was an intended consequence of the autoloader's reconciliation.

> What's the intended behavior of the autoloader?

Personally I think 1 would be antithetical to the goals of the autoloader. Loading an older class seems likely to cause problems when it may have dependencies that will be pulled from a different version of that package. I believe 2 should be the intended behavior. This bug has existed in some form in _every version of the autoloader_, and I wonder how many strange undiagnosable issues it has caused?

It's probably worth saying that fixing this would require changing our manifests and maybe doesn't make sense for a minor version bump. Having the autoloader check for _another_ manifest file in every plugin is continuing to add more disk operations that slow the autoloader down. My preference would probably be that we replace the current manifest files with a single file containing each package, its version, and the classmap, filemap, and PSR-4 map. This would let us reconcile the version of entire modules easily and decrease disk operations to boot.

Reply:
> My preference would probably be that we replace the current manifest files with a single file containing each package, its version, and the classmap, filemap, and PSR-4 map. This would let us reconcile the version of entire modules easily and decrease disk operations to boot.

I mentioned something similar in #17221, towards a different bug. 🙂